File: src/main/java/de/flapdoodle/generics/uniontypes/j/components/FieldFactory.java
package de.flapdoodle.generics.uniontypes.j.components;

import de.flapdoodle.generics.uniontypes.j.framework.Component;
import de.flapdoodle.generics.uniontypes.j.framework.HasValidation;
import de.flapdoodle.generics.uniontypes.j.framework.HasValue;

import java.util.function.Supplier;

public interface FieldFactory {
  <T, C extends Component & HasValue<T> & HasValidation> Supplier<C> factoryFor(Class<T> type);
}

File: src/main/kotlin/de/flapdoodle/generics/uniontypes/k/components/FieldFactory.kt
package de.flapdoodle.generics.uniontypes.k.components

import de.flapdoodle.generics.uniontypes.k.framework.Component
import de.flapdoodle.generics.uniontypes.k.framework.HasValidation
import de.flapdoodle.generics.uniontypes.k.framework.HasValue

interface FieldFactory {
  fun <T, C> factoryFor(type: Class<out T>): () -> C
      where C: Component,
            C: HasValue<T>,
            C: HasValidation
}